File: Other/Phase 2/Review2/binary-search-tree/binary-search-tree.swift
// Binary Search Tree

class TreeNode {
  var val: Int
  var left: TreeNode?
  var right: TreeNode?
  
  init(_ val: Int, _ left: TreeNode? = nil, _ right: TreeNode? = nil) {
    self.val = val
    self.left = left
    self.right = right
  }
}

class BinarySearchTree {
  var root: TreeNode?
  
  init(_ root: TreeNode? = nil) {
    self.root = root
  }
  
  func searchNode(_ value: Int) -> Bool {
    func recursive(_ node: TreeNode?) -> Bool {
      guard let node = node else {
        return false
      }
      
      if value < node.val {
        return recursive(node.left)
      } else if value > node.val {
        return recursive(node.right)
      } else {
        return true
      }
    }
    
    return recursive(root)
  }
  
  func insertNode(_ newValue: Int) -> Bool {
    if root == nil {
      root = TreeNode(newValue)
      return true
    }
    
    func recursion(_ node: TreeNode) -> Bool {
      if newValue < node.val {
        if node.left == nil {
          node.left = TreeNode(newValue)
          return true
        }
        return recursion(node.left!)
      } else if newValue > node.val {
        if node.right == nil {
          node.right = TreeNode(newValue)
          return true
        }
        return recursion(node.right!)
      } else {
        return false
      }
    }
    
    return recursion(root!)
  }
  
  func deleteNode(_ value: Int) -> Bool {
    func recursion(_ node: TreeNode?, _ parent: TreeNode?) -> Bool {
      guard let node = node else {
        return false
      }
      
      if value < node.val {
        return recursion(node.left, node)
      } else if value > node.val {
        return recursion(node.right, node)
      } else {
        if node.left == nil || node.right == nil {
          let child = node.left ?? node.right
          
          if parent == nil {
            root = child
          } else if parent!.left === node {
            parent!.left = child
          } else {
            parent!.right = child
          }
        } else {
          var currentMinNode = node.right
          var parentNode = node
          
          while currentMinNode!.left != nil {
            parentNode = currentMinNode!
            currentMinNode = currentMinNode!.left
          }
          
          node.val = currentMinNode!.val
          if parentNode.left === currentMinNode {
            parentNode.left = currentMinNode!.right
          } else {
            parentNode.right = currentMinNode!.right
          }
        }
        return true
      }
    }
    
    return recursion(root, nil)
  }
  
  func preorder() -> [Int] {
    var res = [Int]()
    
    func recursive(_ node: TreeNode?) {
      guard let node = node else {
        return
      }
      
      res.append(node.val)
      recursive(node.left)
      recursive(node.right)
    }
    
    recursive(root)
    return res
  }
  
  func inorder() -> [Int] {
    var res = [Int]()
    
    func recursive(_ node: TreeNode?) {
      guard let node = node else {
        return
      }
      
      recursive(node.left)
      res.append(node.val)
      recursive(node.right)
    }
    
    recursive(root)
    return res
  }
  
  func postorder() -> [Int] {
    var res = [Int]()
    
    func recursive(_ node: TreeNode?) {
      guard let node = node else {
        return
      }
      
      recursive(node.left)
      recursive(node.right)
      res.append(node.val)
    }
    
    recursive(root)
    return res
  }
}

// test
let tree = BinarySearchTree()
tree.insertNode(8)
tree.insertNode(14)
tree.insertNode(10)
tree.insertNode(13)
tree.insertNode(3)
tree.insertNode(1)
tree.insertNode(6)
tree.insertNode(4)
tree.insertNode(7)
let insertRes1 = tree.preorder()
print(insertRes1)
let searchRes1 = tree.searchNode(14)
print(searchRes1)
let searchRes2 = tree.searchNode(5)
print(searchRes2)

tree.deleteNode(8)
let deleteRes1 = tree.preorder()
print(deleteRes1)

tree.deleteNode(3)
let deleteRes2 = tree.preorder()
print(deleteRes2)

File: Miscellaneous/Other/Review/Stage 1/Review 5/binary-search-tree/binary-search-tree.py
# Binary Search Tree

from typing import List


class TreeNode:

    def __init__(self, val: float, left=None, right=None) -> None:
        self.val: float = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


class BinarySearchTree:
    def __init__(self, node: TreeNode | None) -> None:
        self.root = node

    def searchNode(self, value: float) -> bool:
        def recursive(node: TreeNode | None):
            if not node:
                return False

            if value < node.val:
                return recursive(node.left)
            elif value > node.val:
                return recursive(node.right)
            else:
                return True

        return recursive(self.root)

    def insertNode(self, value: float) -> bool:
        if not self.root:
            self.root = TreeNode(value)
            return True

        def recursion(node: TreeNode):
            if value < node.val:
                if not node.left:
                    node.left = TreeNode(value)
                    return True
                return recursion(node.left)
            elif value > node.val:
                if not node.right:
                    node.right = TreeNode(value)
                    return True
                return recursion(node.right)
            else:
                return False

        return recursion(self.root)

    def deleteNode(self, value: float) -> bool:
        if not self.root:
            return False

        def recursion(node: TreeNode, parent: TreeNode):
            if not node:
                return False

            if value < node.val:
                return recursion(node.left, node)
            elif value > node.val:
                return recursion(node.right, node)
            else:
                if not node.left or not node.right:
                    child = node.left if node.left else node.right
                    if parent == None:
                        self.root = child
                    if parent.left == node:
                        parent.left = child
                    else:
                        parent.right = child
                else:
                    current = node.right
                    parent = node

                    while current.left:
                        parent = current
                        current = current.left

                    node.val = current.val
                    if parent.left == current:
                        parent.left = current.right
                    else:
                        parent.right = current.right

                return True

        return recursion(self.root, None)

    def preorder(self):
        res: List[float] = []

        def recursive(node: TreeNode | None):
            if not node:
                return
            res.append(node.val)
            recursive(node.left)
            recursive(node.right)

        recursive(self.root)
        return res

    def inorder(self):
        res: List[float] = []

        def recursive(node: TreeNode | None):
            if not node:
                return
            recursive(node.left)
            res.append(node.val)
            recursive(node.right)

        recursive(self.root)
        return res

    def postorder(self):
        res: List[float] = []

        def recursive(node: TreeNode | None):
            if not node:
                return
            recursive(node.left)
            recursive(node.right)
            res.append(node.val)

        recursive(self.root)
        return res


# test
tree = BinarySearchTree(None)
tree.insertNode(8)
tree.insertNode(14)
tree.insertNode(10)
tree.insertNode(13)
tree.insertNode(3)
tree.insertNode(1)
tree.insertNode(6)
tree.insertNode(4)
tree.insertNode(7)
insert_res1 = tree.preorder()  # [8, 3, 1, 6, 4, 7, 14, 10, 13]
print(insert_res1)
search_res1 = tree.searchNode(14)  # true
print(search_res1)
search_res2 = tree.searchNode(5)  # false
print(search_res2)

tree.deleteNode(8)
delete_res1 = tree.preorder()  # [10, 3, 1, 6, 4, 7, 14, 13]
print(delete_res1)

tree.deleteNode(3)
delete_res2 = tree.preorder()  # [10, 4, 1, 6, 7, 14, 13]
print(delete_res2)
